src/main/java/ch/njol/skript/effects/EffTransform.java

```diff
@@ -0,0 +1,143 @@
+package ch.njol.skript.effects;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Keywords;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.ExprInput;
+import ch.njol.skript.lang.Effect;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.InputSource;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.skript.variables.Variables;
+import ch.njol.util.Kleenean;
+import ch.njol.util.Pair;
+import ch.njol.util.StringUtils;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+@Name("Transform List")
+@Description({
+	"Transforms (or 'maps') a list's values using a given expression. This is akin to looping over the list and setting " +
+	"each value to a modified version of itself.",
+	"Evaluates the given expression for each element in the list, replacing the original element with the expression's result.",
+	"If the given expression returns a single value, the indices of the list will not change. If the expression returns " +
+	"multiple values, then then indices will be reset as a single index cannot contain multiple values.",
+	"Only variable lists can be transformed with this effect. For other lists, see the transform expression."
+})
+@Examples({
+	"set {_a::*} to 1, 2, and 3",
+	"transform {_a::*} using input * 2",
+	"# {_a::*} is now 2, 4, and 6",
+	"",
+	"# get a list of the sizes of all clans without manually looping",
+	"set {_clan-sizes::*} to indices of {clans::*}",
+	"transform {_clan-sizes::*} using {clans::%input%::size}",
+	"",
+	"# set all existing values of a list to 0:",
+	"transform {_list::*} to 0"
+})
+@Since("INSERT VERSION")
+@Keywords("input")
+public class EffTransform extends Effect implements InputSource {
+
+	static {
+		Skript.registerEffect(EffTransform.class, "(transform|map) %~objects% (using|with) <.+>");
+		if (!ParserInstance.isRegistered(InputData.class))
+			ParserInstance.registerData(InputData.class, InputData::new);
+	}
+
+	private @UnknownNullability Expression<?> mappingExpr;
+	private @UnknownNullability Variable<?> unmappedObjects;
+
+	private final Set<ExprInput<?>> dependentInputs = new HashSet<>();
+
+	private @Nullable Object currentValue;
+	private @UnknownNullability String currentIndex;
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (expressions[0].isSingle() || !(expressions[0] instanceof Variable)) {
```

Comment:
Do you think it would be feasible to support other expressions?
For example:
```
transform the tablist names of all players using "<red>%input%"
```

You could likely use changeInPlace (once merged) to simplify all of this logic

Reply:
Yeah this will get a bit of an overhaul after that's merged

Reply:
After attempts to use it, I think change in place is a bit too limited and unsafe to be used here. I have plans to refactor CIP to make it safer and more versatile, but I don't want to make this PR dependent on that future one. So variables for now, other lists in the future.